### ISSUE TITLE
intellij plugin, problem with agent running commands

### ISSUE THREAD

**Issue by tsobis-generali** [microsoft/copilot-intellij-feedback#679]
IntelliJ IDEA 2025.2.1 (Ultimate Edition)
Build #IU-252.25557.131, built on August 27, 2025
Source revision: ee1e6cb62e111
Runtime version: 21.0.8+1-b1038.68 amd64 (JCEF 122.1.9)
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Toolkit: sun.awt.windows.WToolkit
Windows 11.0
GC: G1 Young Generation, G1 Concurrent GC, G1 Old Generation
Memory: 7168M
Cores: 12
Registry:
  git.recent.checkout.branches.reflog.entries.count=0
Non-Bundled Plugins:
  com.intellij.notebooks.core (252.25557.135)
  com.intellij.classic.ui (252.23892.201)
  Batch Scripts Support (1.0.13)
  org.toml.lang (252.25557.135)
  com.intellij.properties (252.25557.135)
  com.intellij.velocity (252.25557.135)
  CMD Support (1.0.5)
  com.jetbrains.jax.ws (252.23892.360)
  Builder Generator (1.4.0)
  com.intellij.javaee.webSocket (252.23892.360)
  com.intellij.javaee.ejb (252.25557.131)
  intellij.jupyter (252.25557.135)
  com.yiycf.plugins.mavenDependencyHelper (2.5)
  MavenRunHelper (4.30.0-IJ2022.2)
  ch.newinstance.plugin.mavendependencychecker (1.17.0)
  dev.turingcomplete.intellijdevelopertoolsplugins (7.1.0)
  com.intellij.javaee.batch (252.23892.360)
  JRebelPlugin (2025.3.2)
  GlassFish (252.23892.360)
  fish.payara.intellij (2.3.0)
  com.intellij.jsf (252.23892.360)
  com.github.copilot (1.5.53-243)
  org.sonarlint.idea (11.0.0.82343)
  org.jetbrains.security.package-checker (252.25557.135)
Kotlin: 252.25557.131-IJ

Issue: When the Copilot agent proposes and executes shell commands on a user’s machine, it does not reliably detect the operating system before execution and often runs non‑portable commands (e.g., Unix-style on Windows). Even when the OS is specified in the prompt, the agent may still execute a command that fails and does not recognize the failure, proceeding as if the step succeeded.

Impact: This results in frequent no-ops or incorrect changes, especially on Windows hosts, and creates misleading agent state (e.g., subsequent steps assume the prior command succeeded). The lack of robust failure detection increases user friction, reduces trust, and can mask real errors that require remediation.

**Comment by TidyMaze** [comment]
I confirm it happens 50% of the time for me, with gpt 4.1, gpt-5-mini and most of other models.
I feel like Claude models are less impacted by this issue (not sure why, maybe unrelated).

**Comment by shihanMS** [comment]
hi @tsobis-generali, sorry for the late response. We add terminal type as context sent to LLM, and you can use the latest nightly version to see if the issue persist.

**Comment by ghost** [comment]
I was trying to debug where the problem lies. Seems that Copilot has some "environment_info" which says my OS is windows (correct) but assumes my shell is cmd.exe (incorrect). I tried changing this using copilot instructions, settings in IDE etc, but nothing worked. 

It even entered some loop at one point, getting confused as I insisted my shell is bash, it should use only bash, it should never use cmd. He then went on in a long text about how my information and "his" information provided in "environment_info" and apparently system prompt block him from following my instructions and explicitely force him to use the assumed shell. 

I get that you may want to do some assumptions to reduce the manual config on the developer side, but please make those settings always visible and always configurable. 

At one point I was really thinking on giving up as it is impossible to work with something that refuses to follow instructions and forces itself to assume wrong information blocking itself.

I will try the nightly and hopefully it is fixed...

**Comment by shihanMS** [comment]
hi @IvanCarapovic, if no luck, could u pls share `idea.log` here? You can refer to the following guide for instructions on enabling debug logging: https://github.com/microsoft/copilot-intellij-feedback/wiki/Enable-debug-logging

If you're not comfortable sharing the full log, you can instead search for the line: "Extracted shell `shellType` from run_in_terminal tool description" and check whether the extracted `shellType` appears to be correct.
